### PR TITLE
Add compiler gym actor_critic example.

### PR DESCRIPTION
**Pull request by broune**
Add compiler gym actor-critic example.

This example is simplified by only training for a single benchmark at a time, using a subset of 15 LLVM compiler passes and using a fixed episode length.

Here is a paste with a bunch of data from just playing around with the parameters:

  https://pastebin.com/raw/HgS1GCEe

If run with no parameters, it takes 11 seconds to find a good configuration on my macbook. Likely this could be made much faster but this is just supposed to be a fairly simple example, so I think that is fine.

Example output from running python actor_critic.py :

Seed: 101
Episode length: 5
Exploration: 0.00%
Learning rate: 0.008
Action space: ['-simplifycfg', '-reg2mem', '-early-cse-memssa', '-gvn-hoist', '-gvn', '-instsimplify', '-instcombine', '-jump-threading', '-loop-extract', '-loop-reduce', '-loop-versioning', '-newgvn', '-mem2reg', '-sroa', '-structurizecfg']
Episode 1	Last reward: 52.00	Avg reward: 52.00	Best reward: 52.00	Last loss: 0.410147	Avg loss: 0.410147	
Episode 100	Last reward: 97.00	Avg reward: 118.10	Best reward: 153.00	Last loss: -0.045142	Avg loss: 0.432208	
Episode 200	Last reward: 140.00	Avg reward: 130.32	Best reward: 153.00	Last loss: 3.115863	Avg loss: 0.219633	
Episode 300	Last reward: 131.00	Avg reward: 135.27	Best reward: 153.00	Last loss: -0.087667	Avg loss: -0.016292	
Episode 400	Last reward: 143.00	Avg reward: 140.41	Best reward: 153.00	Last loss: -0.556506	Avg loss: 0.031324	
Episode 500	Last reward: 152.00	Avg reward: 145.52	Best reward: 153.00	Last loss: 0.899966	Avg loss: 0.032300	
Episode 600	Last reward: 152.00	Avg reward: 150.30	Best reward: 153.00	Last loss: 0.076278	Avg loss: -0.001032	
Episode 700	Last reward: 152.00	Avg reward: 151.88	Best reward: 153.00	Last loss: -0.059516	Avg loss: -0.024967	
Episode 800	Last reward: 153.00	Avg reward: 152.25	Best reward: 153.00	Last loss: -0.138806	Avg loss: -0.066543	
Episode 900	Last reward: 153.00	Avg reward: 152.70	Best reward: 153.00	Last loss: -0.001581	Avg loss: -0.025742	
Episode 1000	Last reward: 153.00	Avg reward: 152.97	Best reward: 153.00	Last loss: 0.000665	Avg loss: 0.016727	

Final performance (avg reward): 152.97
Final avg reward versus own best: -0.03

### CODE REVIEW

**Comment by ChrisCummins** [comment]
Hi @broune, fantastic! The implementation is straightforward enough and I really appreciate the good comments.

The reason you had to compute reward yourself is because I renamed the `eager_reward_space` property to [env.reward_space](https://facebookresearch.github.io/CompilerGym/compiler_gym/envs.html#compiler_gym.envs.CompilerEnv.reward_space). I made a patch to update the property name and add it as a flag, you can apply it by copying and pasting this into the terminal:

```patch
git apply - <<EOF
diff --git a/examples/actor_critic.py b/examples/actor_critic.py
index 5c8d7bf..7736a81 100644
--- a/examples/actor_critic.py
+++ b/examples/actor_critic.py
@@ -57,6 +57,7 @@ flags.DEFINE_list(
     ],
     "List of optimizatins to explore.",
 )
+flags.DEFINE_string("reward", "IrInstructionCount", "The reward function to optimize.")
 flags.DEFINE_string("benchmark", "cBench-v0/dijkstra", "Benchmark to optimize.")
 flags.DEFINE_integer("episode_len", 5, "Number of transitions per episode.")
 flags.DEFINE_integer("hidden_size", 64, "Latent vector size.")
@@ -116,10 +117,9 @@ class CustomEnv:
     """

     def __init__(self):
-        self._env = gym.make("llvm-v0")
+        self._env = gym.make("llvm-v0", reward_space=FLAGS.reward)
         try:
             self._env.require_dataset("cBench-v0")
-            self._env.eager_reward_space = "IrInstructionCount"
             self._env.reset(benchmark=FLAGS.benchmark)

             # Project onto the subset of transformations that have
@@ -147,18 +147,12 @@ class CustomEnv:

         # Update environment
         _, reward, done, info = self._env.step(self._actions[action])
-        assert reward is None  # This is supposed to be the reward, but isn't
         done = done or self._steps_taken == FLAGS.episode_len

-        size = float(self._env.observation["IrInstructionCount"])
-        reward = self.prior_size - size
-        self.prior_size = size
-
         return self._state, reward, done, info

     def reset(self):
         self._env.reset()
-        self.prior_size = float(self._env.observation["IrInstructionCount"])
         self._steps_taken = 0
         self._state = np.zeros(
             (FLAGS.episode_len - 1, len(self.action_space)), dtype=np.int32
@@ -386,6 +380,8 @@ def main(argv):
         print(f"Episode length: {FLAGS.episode_len}")
         print(f"Exploration: {FLAGS.exploration:.2%}")
         print(f"Learning rate: {FLAGS.learning_rate}")
+        print(f"Reward: {FLAGS.reward}")
+        print(f"Benchmark: {FLAGS.benchmark}")
         print(f"Action space: {env.action_space}")

         if FLAGS.iterations == 1:
EOF
```

If we use the [IrInstructionCountOz](https://facebookresearch.github.io/CompilerGym/llvm/index.html#reward-spaces) reward we see that the actor critic does really quite well given only 5 passes:

```
$ python actor_critic.py --reward=IrInstructionCountOz
Seed: 101
Episode length: 5
Exploration: 0.00%
Learning rate: 0.008
Reward: IrInstructionCountOz
Benchmark: cBench-v0/dijkstra
Action space: ['-simplifycfg', '-reg2mem', '-early-cse-memssa', '-gvn-hoist', '-gvn', '-instsimplify', '-instcombine', '-jump-threading', '-loop-extract', '-loop-reduce', '-loop-versioning', '-newgvn', '-mem2reg', '-sroa', '-structurizecfg']
Episode 1	Last reward: 0.34	Avg reward: 0.34	Best reward: 0.34	Last loss: 0.410146	Avg loss: 0.410146
Episode 100	Last reward: 0.63	Avg reward: 0.76	Best reward: 0.99	Last loss: -0.045142	Avg loss: 0.432207
Episode 200	Last reward: 0.90	Avg reward: 0.84	Best reward: 0.99	Last loss: 3.115863	Avg loss: 0.219632
Episode 300	Last reward: 0.85	Avg reward: 0.87	Best reward: 0.99	Last loss: -0.087670	Avg loss: -0.016293
Episode 400	Last reward: 0.92	Avg reward: 0.91	Best reward: 0.99	Last loss: -0.556499	Avg loss: 0.031323
Episode 500	Last reward: 0.98	Avg reward: 0.94	Best reward: 0.99	Last loss: 0.899853	Avg loss: 0.032295
Episode 600	Last reward: 0.98	Avg reward: 0.97	Best reward: 0.99	Last loss: 0.076221	Avg loss: -0.001041
Episode 700	Last reward: 0.98	Avg reward: 0.98	Best reward: 0.99	Last loss: -0.124386	Avg loss: -0.013512
Episode 800	Last reward: 0.98	Avg reward: 0.98	Best reward: 0.99	Last loss: -0.064390	Avg loss: -0.057069
Episode 900	Last reward: 0.99	Avg reward: 0.99	Best reward: 0.99	Last loss: -0.001138	Avg loss: -0.005302
Episode 1000	Last reward: 0.99	Avg reward: 0.99	Best reward: 0.99	Last loss: 0.001477	Avg loss: -0.008119
```

Instruction count reduction is 153 against `clang -Oz`'s 155:

```py
ic_base = env._env.observation["IrInstructionCountO0"]
ic_oz = env._env.observation["IrInstructionCountOz"]
ic = env._env.observation["IrInstructionCount"]
print("Instruction count reductions: Learned policy:", int(ic_base - ic), "-Oz:", int(ic_base - ic_oz))
# Instruction count reductions: Learned policy: 153 -Oz: 155
```

With the reward patch above this looks good to merge. The other requirement is a smoke test, which I am happy to add after merging.

Cheers,
Chris

**Comment by broune** [comment]
Thanks. Added the configurable reward space flag.

File: examples/actor_critic.py [hunk 1]
```diff
@@ -0,0 +1,413 @@
...
+    probs, state_value = model(state)
+
+    # Create a probability distribution where the probability of
+    # action i is probs[i].
+    m = Categorical(probs)
+
+    # Sample an action using the distribution, or pick an action
+    # uniformly at random if in an exploration mode.
+    if random.random() < exploration_rate:
+        action = torch.tensor(random.randrange(0, len(probs)))
```
Comment (soumith) [inline examples/actor_critic.py:228]: fyi, i think you can use `torch.randint` here if you want, though it's probably an overkill and a bit of overhead for drawing a single number

File: examples/actor_critic.py [hunk 2]
```diff
@@ -0,0 +1,413 @@
...
+        self.moving_mean = MovingExponentialAverage(FLAGS.mean_smoothing)
+        self.moving_std = MovingExponentialAverage(FLAGS.std_smoothing)
+
+    def forward(self, x):
+        """Forward of both actor and critic"""
+        # Initial layer maps the sequence of one-hot vectors into a
+        # vector of the hidden size. Next layers stay with the same
+        # size and use residual connections.
+        x = F.relu(self.affine1(x))
+        x = x.add(F.relu(self.affine2(x)))
```
Comment (soumith) [inline examples/actor_critic.py:199]: i've mostly seen residual connections used when the branch has multiple layers. but if this works in practice, great :)

File: examples/actor_critic.py [hunk 3]
```diff
@@ -0,0 +1,413 @@
...
+    # use a discount factor as the episode length is fixed and not
+    # very long, but if we had used one, it would appear here.
+    for r in model.rewards[::-1]:
+        R += r
+        returns.insert(0, R)
+
+    # Update the moving averages for mean and standard deviation and
+    # use that to normalize the input.
+    returns = torch.tensor(returns)
+    model.moving_mean.next(returns.mean())
```
Comment (soumith) [inline examples/actor_critic.py:262]: is the motivation to use moving averages primarily because the episode length is short?

**Review by soumith — APPROVED** [review]
nicely done. lgtm!

File: examples/actor_critic.py (hunk 2 above)
Comment (broune) [reply examples/actor_critic.py:199]: I just figured I'd try it. I ran an experiment in the linked pastebin that shows that performance is substantially worse without the residual connections, so I left them in. (Though just removing the additional layers is also supported by that data)

File: examples/actor_critic.py (hunk 3 above)
Comment (broune) [reply examples/actor_critic.py:262]: The justification for using it is the experiments in the linked pastebin that show that it works better. That may be related to short episode length, and the benefit may fall off for longer episode lengths, I don't know since I didn't try that.

My motivation for trying this in the first place is that data normalization for the input data like this is normally done across the entire dataset at once, so N=everything, which we can't do here since we're generating and training on one data point at a time, so it would be done N=1 as in the official PT example, but N=1 is very different from N=everything and I was thinking that this might make each data point look quite different to the model in a way that it can't sort out. Another way of looking at it is that this is like batchnorm for minibatch=1, where it is known that minibatch should ideally be more like 32 for best results. Doing the normalization even for N=1 is dramatically better than not doing any normalization, though a moving mean works even better.
